pkg/controller/direct/run/job_controller.go

```diff
@@ -208,39 +201,21 @@ func (a *JobAdapter) Update(ctx context.Context, updateOp *directbase.UpdateOper
 	log := klog.FromContext(ctx)
 	log.V(2).Info("updating Job", "name", a.id)
 
-	specHash, err := common.HashProto(a.desired)
-	if err != nil {
-		return fmt.Errorf("calculating spec hash: %w", err)
-	}
-	gcpHash, err := common.HashProto(a.actual)
+	currentCookie, err := common.NewCookie(a.desired, a.actual)
 	if err != nil {
-		return fmt.Errorf("calculating gcp hash: %w", err)
+		return err
 	}
-
-	if a.lastModifiedCookie != nil {
-		cookie, err := common.ParseCookie(direct.ValueOf(a.lastModifiedCookie))
-		if err != nil {
-			log.V(2).Info("could not parse cookie, forcing update", "cookie", direct.ValueOf(a.lastModifiedCookie), "err", err)
-		} else {
-			if cookie.SpecHash == specHash && cookie.GCPHash == gcpHash {
-				log.V(2).Info("resource is up to date", "name", a.id)
-				// Fast path: Update status with observed state and return.
-				// The status update is important to update conditions and observedGeneration.
-				status := &krm.RunJobStatus{}
-				mapCtx := &direct.MapContext{}
-				status.ObservedState = RunJobObservedState_FromProto(mapCtx, a.actual)
-				if mapCtx.Err() != nil {
-					return mapCtx.Err()
-				}
-				status.LastModifiedCookie = a.lastModifiedCookie // Preserve cookie
-				return updateOp.UpdateStatus(ctx, status, nil)
-			}
-		}
+	if currentCookie.Equal(a.lastModifiedCookie) {
+		log.V(2).Info("resource is up to date", "name", a.id)
+		return a.updateStatus(ctx, a.actual, updateOp)
 	}
 
-	a.desired.Name = a.actual.Name
+	// We need to set the name for the update request, but we don't want to modify a.desired
```

Comment:
We should be generating a structured diff and reporting it when we decide we need to do an actual update.

Reply:
Good call!